### PR TITLE
Index factory

### PR DESCRIPTION
**Pull request by meiravgri**
Index factory purpose is to manage creation and info index functions.

**Describe the changes in the pull request**

## Motivation 
As we have and will elaborte to support more index configuration, 
we decide to create an independent class that will manage "logistics".

The functionalities currently moved from the index classes to the factory classes are:
* Allocation new index
* Supply memory info like initial size or element size.

**Which issues this PR fixes**
1. Dependency of the base class on its inherited class

**Main objects this PR modified**
1. Indexes internal factory and memory estimation functions

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

File: src/VecSim/algorithms/brute_force/brute_force_factory.cpp [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+    if (params->initialCapacity) {
+        est += params->initialCapacity * sizeof(labelType) + sizeof(size_t);
+    }
+
+    return est;
+}
+
+size_t BruteForceFactory::EstimateElementSize(const BFParams *params) {
+    return params->dim * VecSimType_sizeof(params->type) + sizeof(idType);
```
Comment (GuyAv46) [inline src/VecSim/algorithms/brute_force/brute_force_factory.cpp:30]:
```suggestion
    return params->dim * VecSimType_sizeof(params->type) + sizeof(labelType);
```

File: src/VecSim/algorithms/hnsw/hnsw_factory.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    est += sizeof(void *) * params->initialCapacity + sizeof(size_t); // link lists (for levels > 0)
+    est += sizeof(size_t) * params->initialCapacity + sizeof(size_t); // element level
+    est += sizeof(size_t) * params->initialCapacity +
+           sizeof(size_t); // Labels lookup hash table buckets.
+
+    size_t size_links_level0 =
+        sizeof(linklistsizeint) + params->M * 2 * sizeof(idType) + sizeof(void *);
+    size_t size_total_data_per_element =
+        size_links_level0 + params->dim * VecSimType_sizeof(params->type) + sizeof(labelType);
+    est += params->initialCapacity * size_total_data_per_element + sizeof(size_t);
```
Comment (GuyAv46) [inline src/VecSim/algorithms/hnsw/hnsw_factory.cpp:37]: Shouldn't this all be calculated only if `params->initialCapacity` is not 0?
Also, we can have `idType` and `labelType` here now.

File: src/VecSim/algorithms/hnsw/hnsw_factory.cpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+    return est;
+}
+
+size_t HNSWFactory::EstimateElementMemory(const HNSWParams *params) {
```
Comment (GuyAv46) [inline src/VecSim/algorithms/hnsw/hnsw_factory.cpp:42]: please have the names consistent.

File: src/VecSim/vec_sim.cpp [hunk 4]
```diff
@@ -115,9 +117,9 @@ extern "C" double VecSimIndex_GetDistanceFrom(VecSimIndex *index, size_t id, con
 extern "C" size_t VecSimIndex_EstimateElementSize(const VecSimParams *params) {
     switch (params->algo) {
     case VecSimAlgo_HNSWLIB:
-        return HNSWIndex::estimateElementMemory(&params->hnswParams);
+        return HNSWFactory::EstimateElementMemory(&params->hnswParams);
```
Comment (GuyAv46) [inline src/VecSim/vec_sim.cpp:120]: please have the names consistent.

**Review by GuyAv46 — COMMENTED** [review]
Nicely done
few comments

File: src/VecSim/algorithms/hnsw/hnsw_factory.cpp (hunk 2 above)
Comment (meiravgri) [reply src/VecSim/algorithms/hnsw/hnsw_factory.cpp:37]: no, there is additional overhead regardless of the initial size

**Review by GuyAv46 — APPROVED** [review]
LGTM :)

**Review by DvirDukhan — APPROVED** [review]

**Pull request merged**